src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -1238,6 +1249,34 @@ internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, r
             }
         }
 
+        /// <summary>
+        /// Returns true if the type is generic or non-generic custom task-like type due to the
+        /// [AsyncMethodBuilder(typeof(B))] attribute. It returns the "B".
+        /// </summary>
+        /// <param name="builderArgument">The argument to the <see cref="System.Runtime.CompilerServices.AsyncMethodBuilderAttribute"/> or null if not a custom task</param>
+        /// <returns>True if this type is a custom task type</returns>
+        /// <remarks>
+        /// The definition of "custom task-like type" is one that has an [AsyncMethodBuilder(typeof(B))] attribute,
+        /// no more, no less. Validation of builder type B is left for elsewhere. This method returns B
+        /// without validation of any kind.
+        /// 
+        /// This specialization retrieves B from the early decoded attributed data: this prevents binding loops 
+        /// when we're binding attributes on a type and need to know during binding if the same type is task-like
+        /// </remarks>
+        public override bool IsCustomTaskType(out object builderArgument)
```

Comment:
>public override bool IsCustomTaskType(out object builderArgument) [](start = 8, length = 65)

Is this override is sufficient to handle generic types as well? Please add a test that involves a constructed generic type.